docs/proposals/synchronous-garbage-collection.md

```diff
@@ -0,0 +1,140 @@
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Overview
+
+Users of the server-side garbage collection need to determine if the garbage collection is done. For example:
+* Currently `kubectl delete rc` blocks until all the pods are terminating. To convert to use server-side garbage collection, kubectl has to be able to determine if the garbage collection is done.
+* [#19701](https://github.com/kubernetes/kubernetes/issues/19701#issuecomment-236997077) is a use case where the user needs to wait for all service dependencies garbage collected and their names released, before she recreates the dependencies.
+
+We define the garbage collection as "done" when all the dependents are deleted from the key-value store, rather than merely in the terminating state. There are two reasons: *i)* for `Pod`s, the most usual garbage, only when they are deleted from the key-value store, we know kubelet has released their resources; *ii)* some users need to recreate objects with the same names, they need to wait for the old objects to be deleted from the key-value store. (This limitation is because we index objects by their names in the key-value store today.)
+
+Synchronous Garbage Collection is a best-effort (see [unhandled cases](#unhandled-cases)) mechanism that allows user to determine if the garbage collection is done: after the API server receives a deletion request of an owning object, the object keeps existing in the key-value store until all its dependents are deleted from the key-value store by the garbage collector.
+
+Tracking issue: https://github.com/kubernetes/kubernetes/issues/29891
+
+# Required code modification
+
+We need to make changes in the API, the API Server, and the garbage collector to support synchronous garbage collection.
+
+## API changes
+
+```go
+DeleteOptions {
+  …
+  // If SynchronousGarbageCollection is set, the object will not be deleted immediately. Instead, a CollectingGarbage finalizer will be placed on the object. The garbage collector will remove the finalizer from the object when all depdendents are deleted.
+  // SynchronousGarbageCollection and OrphanDependents are exclusive.
+  // SynchronousGarbageCollection defaults to false.
+  // SynchronousGarbageCollection is cascading, i.e., the object’s dependents will be deleted with the same SynchronousGarbageCollection.
+  SynchronousGarbageCollection *bool
+}
+```
+
+We will introduce a new standard finalizer: const GCFinalizer string = “CollectingGarbage”
+
+## Components changes
+
+### API Server
+
+Delete() function needs to check the `DeleteOptions.SynchronousGarbageCollection`.
+
+* The request is rejected with 400 if both `DeleteOptions.SynchronousGarbageCollection` and `DeleteOptions.OrphanDependents` are true (or default to true, see [DefaultGarbageCollectionPolicy](https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/registry/generic/registry/store.go#L500)).
```

Comment:
If either one is set, I think we should make the most generous assumption possible about the other one. e.g. if you specify sync=true and don't specify orphan at all, we should assume you mean orphan=false, even if the resource would ordinarily default to orphaning.